rfd/0031-dynamic-app-db-registration.md

```diff
@@ -0,0 +1,355 @@
+components of the system i.e. when a new app or database service agent joins
+the cluster, it needs to present a valid auth token. Once the service has
+registered, updating its configuration to connect new web apps or databases
+does not require re-registering the service with the cluster.
+
+Dynamic app/database registration plugs into this model - when a new app or
+database resource is created, it is picked up by a service that is already
+part of the cluster. Managing app/database resources themselves is subject to
+RBAC checks so only users with appropriate permissions to `app_server` and
+`db_server` resources are allowed to create/remove them.
+
+## API changes
+
+Teleport API client currently provides the following methods for managing
+app/database resources (copied from [api/client/client.go](https://github.com/gravitational/teleport/blob/v7.0.0-dev.9/api/client/client.go)):
+
+```go
+GetAppServers(ctx context.Context, namespace string, skipValidation bool) ([]types.Server, error)
+UpsertAppServer(ctx context.Context, server types.Server) (*types.KeepAlive, error)
+DeleteAppServer(ctx context.Context, namespace, name string) error
+```
+
+```go
+GetDatabaseServers(ctx context.Context, namespace string, skipValidation bool) ([]types.DatabaseServer, error)
+UpsertDatabaseServer(ctx context.Context, server types.DatabaseServer) (*types.KeepAlive, error)
+DeleteDatabaseServer(ctx context.Context, namespace, hostID, name string) error
+```
+
+These methods were designed for internal usage by Teleport components e.g. when
+app/database agent registers with the cluster or for RBAC checks. As such, they
+are quite awkward for API client users. Changes to make them more API friendly:
+
+* If we decide to settle on `app` and `db` terminology (vs current `app_server`
+  and `db_server`), name them appropriately like `GetApps` or `DeleteDatabase`.
+* Switch to `Create` and `Update` methods instead of `Upsert` anti-pattern.
+* Remove extra arguments that either aren't used or don't make sense for users
+  such as `namespace` (always "default"), `skipValidation`, `hostID`, etc.
+* Add methods for querying individual apps/databases by names which is useful
+  for Terraform provider.
+
+Cleaned up version of the apps API (using Go syntax instead of `.proto` for
+brevity):
+
+```go
+GetApps(ctx context.Context) ([]types.AppServer, error)
+GetApp(ctx context.Context, name string) (types.AppServer, error)
+CreateApp(ctx context.Context, app types.AppServer) error
+UpdateApp(ctx context.Context, app types.AppServer) error
+DeleteApp(ctx context.Context, name string) error
+```
+
+```go
+GetDatabases(ctx context.Context) ([]types.DatabaseServer, error)
+GetDatabase(ctx context.Context, name string) (types.DatabaseServer, error)
+CreateDatabase(ctx context.Context, db types.DatabaseServer) error
+UpdateDatabase(ctx context.Context, db types.DatabaseServer) error
+DeleteDatabase(ctx context.Context, name string) error
+```
+
+## Kubernetes
```

Comment:
FYI, I was tracking various similar ideas in https://github.com/gravitational/teleport/issues/4832
We can leave all deeper k8s integration to that issue and a separate RFD.